crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -226,6 +239,8 @@ impl<'db> SemanticIndexBuilder<'db> {
             DefinitionCategory::Declaration => use_def.record_declaration(symbol, definition),
             DefinitionCategory::Binding => use_def.record_binding(symbol, definition),
         }
+        let snapshot = self.flow_snapshot();
```

Comment:
Oh, pretty sure this is the reason for the perf regression. I'm now taking a snapshot for every definition, even if it doesn't occur inside an exception handler 🤦

Will fix tomorrow.

Reply:
This diagnosis was correct. https://github.com/astral-sh/ruff/pull/13729/commits/24d139891efc6030ef0f3ec2eebd1999c2e6e831 fixed nearly all of the performance regression; it's down to 1-2% now.